server/src/main/scala/services/OptimiserPlus.scala

```diff
@@ -0,0 +1,387 @@
+                  cost: IndexedSeq[Int] => Cost,
+                  xmin: IndexedSeq[Int],
+                  xmax: IndexedSeq[Int],
+                  concavityLimit: Int): Iterable[Int] = {
+    val desks = startingX.to[mutable.IndexedSeq]
+    var incumbent = startingX
+    val minutes = desks.length
+    var bestSoFar = cost(incumbent.toIndexedSeq).totalPenalty
+    val candidates = (0 until minutes)
+      .map(i => neighbouringPoints(startingX(i), xmin(i), xmax(i)))
+      .to[mutable.IndexedSeq]
+
+    var cursor = minutes - 1
+
+    while (cursor >= 0) {
+      while (candidates(cursor).nonEmpty) {
+        desks(cursor) = candidates(cursor).head
+        candidates(cursor) = candidates(cursor).drop(1)
+
+        val trialPenalty = cost(desks.toIndexedSeq).totalPenalty
+
+        if (trialPenalty > bestSoFar + concavityLimit) {
+          if (desks(cursor) > incumbent(cursor)) {
+            candidates(cursor) = candidates(cursor).filter(_ < desks(cursor))
+          } else {
+            candidates(cursor) = candidates(cursor).filter(_ > desks(cursor))
+          }
+        } else {
+          if (trialPenalty < bestSoFar) {
+            incumbent = desks.toIndexedSeq
+            bestSoFar = trialPenalty
+          }
+          if (cursor < minutes - 1) cursor = cursor + 1
+        }
+      }
+      candidates(cursor) = neighbouringPoints(incumbent(cursor), xmin(cursor), xmax(cursor))
+      desks(cursor) = incumbent(cursor)
+      cursor = cursor - 1
+    }
+    desks
+  }
+
+  def tryOptimiseWin(work: IndexedSeq[Double],
+                     minDesks: IndexedSeq[Int],
+                     maxDesks: IndexedSeq[Int],
+                     sla: Int,
+                     weightChurn: Double,
+                     weightPax: Double,
+                     weightStaff: Double,
+                     weightSla: Double,
+                     processors: WorkloadProcessorsLike): Try[IndexedSeq[Int]] = {
+    if (work.length != minDesks.length) {
+      Failure(new Exception(s"work & minDesks are not equal length: ${work.length} vs ${minDesks.length}"))
+    } else if (work.length != maxDesks.length) {
+      Failure(new Exception(s"work & maxDesks are not equal length: ${work.length} vs ${maxDesks.length}"))
+    } else Try {
+      val blockWidth = 15
+      val concavityLimit = 30
+      val winStep = 60
+      val smoothingWidth = blockWidth
+      val winWidth = List(90, work.length).min
+
+      var winStart = 0
+      var winStop = winWidth
+      var qStart = IndexedSeq(0d)
+      var churnStart = 0
+
+      val desks = blockMean(runningAverage(work, smoothingWidth, processors), blockWidth)
+        .zip(maxDesks)
+        .map {
+          case (d, max) => List(d, max).min
+        }
+        .zip(minDesks)
+        .map {
+          case (d, min) => List(d, min).max
+        }.to[mutable.IndexedSeq]
+
+      def myCost(costWork: IndexedSeq[Double], costQStart: IndexedSeq[Double], costChurnStart: Int)
```

Comment:
It's called `myCost` because I lazily didn't spend much time figuring out what it was doing when I translated the R code a couple of years ago. It's not `workCost` because it's actually taking into account many different factors that have a cost associated with them